hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,266 @@
+            if (ndim < 3) {
+               static_cast<TH1*>(h_merged)->Add(static_cast<TH1*>(h));
+            } else {
+               static_cast<THnBase*>(h_merged)->Add(static_cast<THnBase*>(h));
+            }
+
+            delete h;
+         } else {
+            h_merged = h;
+         }
+      }
+
+      return h_merged;
+   }
+
+   /// Retrieve a histogram from a file.
+   ///
+   /// \param fileName path of the file to read.
+   THnBase* ReadHistogram(const char* fileName)
+   {
+      TFile* f = TFile::Open(fileName);
+      f->cd();
+      THnBase* hs = nullptr;
+      f->GetObject(fName, hs);
+      delete f;
+
+      return hs;
+   }
+
+   /// Copy the properties of all axes a histogram.
+   ///
+   /// \param hs histogram whose axes should be updated
+   void SetupAxes(THnBase& hs)
+   {
+      const Int_t naxes = fAxes.GetEntries();
+      for (Int_t i = 0; i < naxes; ++i) {
+         const TAxis* ax_ref = static_cast<TAxis*>(fAxes[i]);
+         TAxis* ax = hs.GetAxis(i);
+         ax_ref->Copy(*ax);
+      }
+   }
+
+   /// Ensure a histogram has axes similar to the ones we expect.
+   ///
+   /// \param h histogram to verify
+   /// \param axes expected set of axes
+   static bool CheckConsistency(const THnBase& h, const TObjArray& axes)
+   {
+      // We would prefer to directly use `TH1::CheckEqualAxes` here;
+      // however it is protected so we inherit the parts we care about.
+      // FIXME(bbannier): It appears that functionality like `TH1::CheckEqualAxes` could
+      // just as well live in `TAxis` so that anyone using axes could make use of it.
+      const Int_t naxes = h.GetNdimensions();
+      const Int_t naxes2 = axes.GetEntries();
+      if (naxes != naxes2) return false;
+
+      for (Int_t i = 0; i < naxes; ++i) {
+         const TAxis* ax1 = h.GetAxis(i);
+         const TAxis* ax2 = static_cast<TAxis*>(axes[i]);
+
+         if (ax1->GetNbins() != ax2->GetNbins()) {
+            return false;
+         }
+
+         // Copied from `TH1::CheckAxisLimits.`
+         if (!TMath::AreEqualRel(ax1->GetXmin(), ax2->GetXmin(), 1.E-12) ||
+             !TMath::AreEqualRel(ax1->GetXmax(), ax2->GetXmax(), 1.E-12)) {
+            return false;
+         }
+
+         // Copied from `TH1::CheckBinLimits`.
+         const TArrayD* h1Array = ax1->GetXbins();
+         const TArrayD* h2Array = ax2->GetXbins();
+         Int_t fN = h1Array->fN;
+         if (fN != 0) {
+            if (h2Array->fN != fN) {
+               return false;
+            } else {
+               for (int ibin = 0; ibin < fN; ++ibin) {
+                  if (!TMath::AreEqualRel(h1Array->GetAt(ibin), h2Array->GetAt(ibin), 1E-10)) {
+                     return false;
+                  }
+               }
+            }
+         }
+
+         // We ignore checking for equal bin labels here. A check
+         // for that is implemented in `TH1::CheckBinLabels`.
+      }
+
+      return true;
+   }
+
+   ClassDef(THnChain, 0);
```

Comment:
Right now this class does not own the `TAxis*` it stores internally, so it would require some additional adjustment for when the `THnChain` is restored from file. What method would I need to overload for that?